Sources/SwiftKafka/KafkaProducer.swift

```diff
@@ -14,21 +14,183 @@
+
+    let _internal = NIOAsyncSequenceProducer.makeSequence(
+        of: Element.self,
+        backPressureStrategy: NoBackPressure(),
+        delegate: NoDelegate()
+    )
+
+    func produceMessage(_ message: KafkaAckedMessage) {
+        _ = self._internal.source.yield(message)
+    }
+
+    public func makeAsyncIterator() -> AsyncIterator {
+        self._internal.sequence.makeAsyncIterator()
+    }
+}
 
 /// Send messages to the Kafka cluster.
+/// Please make sure to explicitly call ``shutdownGracefully()`` when the `KafkaProducer` is not used anymore.
 /// - Note: When messages get published to a non-existent topic, a new topic is created using the ``KafkaTopicConfig``
 /// configuration object (only works if server has `auto.create.topics.enable` property set).
-public struct KafkaProducer {
-    private let client: KafkaClient
+public actor KafkaProducer {
+    /// `AsyncSequence` that returns all ``KafkaProducerMessage`` objects that have been
+    /// acknowledged by the Kafka cluster.
+    public nonisolated let acknowledgements = AcknowledgedMessagesAsyncSequence()
+
+    private var messageIDCounter: UInt = 0
+    private var config: KafkaConfig
     private let topicConfig: KafkaTopicConfig
+    private let logger: Logger
+    private var topicHandles: [String: OpaquePointer]
 
-    // Preliminary implementation
+    // Implicitly unwrapped optional because we must pass self to our config before initializing client etc.
+    private var referenceToSelf: UnsafeMutableRawPointer!
+    private var client: KafkaClient!
+    private var kafkaHandle: OpaquePointer!
+    private var pollTask: Task<Void, Never>!
+
+    /// Initialize a new `KafkaProducer`.
+    /// - Parameter config: The ``KafkaConfig`` for configuring the `KafkaProducer`.
+    /// - Parameter topicConfig: The ``KafkaTopicConfig`` used for newly created topics.
+    /// - Parameter logger: A logger.
     public init(
         config: KafkaConfig = KafkaConfig(),
         topicConfig: KafkaTopicConfig = KafkaTopicConfig(),
         logger: Logger
-    ) throws {
-        self.client = try KafkaClient(type: .producer, config: config, logger: logger)
+    ) async throws {
+        self.config = config
         self.topicConfig = topicConfig
+        self.logger = logger
+        self.topicHandles = [:]
+
+        self.referenceToSelf = Unmanaged.passRetained(self).toOpaque()
+        self.config.setDeliveryReportCallback(callback: self.deliveryReportCallback)
+        self.config.setCallbackOpaque(opaque: self.referenceToSelf)
+
+        self.client = try KafkaClient(type: .producer, config: self.config, logger: self.logger)
+        self.kafkaHandle = self.client.kafkaHandle
+
+        // Poll Kafka every millisecond
+        self.pollTask = Task { [kafkaHandle] in
+            while !Task.isCancelled {
+                rd_kafka_poll(kafkaHandle, 0)
+                try? await Task.sleep(nanoseconds: 1_000_000)
```

Comment:
Can we use the `Clock` API here instead of the nanosecond one?

Reply:
(After chatting, we agreed on waiting until `Swift 5.7` is officially out)